Cython/Compiler/Symtab.py

```diff
@@ -2517,8 +2517,11 @@ def declare_cfunction(self, name, type, pos,
                 if entry.is_final_cmethod and entry.is_inherited:
                     error(pos, "Overriding final methods is not allowed")
                 elif type.same_c_signature_as(entry.type, as_cmethod = 1) and type.nogil == entry.type.nogil:
-                    # Fix with_gil vs nogil.
-                    entry.type = entry.type.with_with_gil(type.with_gil)
+                    if self.parent_type.is_final_type:
+                        entry = self.add_cfunction(name, type, pos, cname, visibility='ignore', modifiers=modifiers)
+                    else:
+                        # Fix with_gil vs nogil.
+                        entry.type = entry.type.with_with_gil(type.with_gil)
```

Comment:
I just wonder if something to test this should be included in the test? It isn't obvious to me whether this should be inside an `else` or not.

Reply:
This whole thing confuses me as well. I have not idea what to exactly test for. I just assumed that if a new final method is being added to a scopy, the input type information and its gil properties are already properly set, therefore you just create a new entry. 

The else branch is just leaving things as they were before, without messing with them in any way. I'm just trying to fix the underlying issue the right way, not with a low-quality temporary workaround as before, while minimizing the risk of breaking something. It looks really weird that this else branch would mess in such a way with `entry.type` in for an already-declared entry that comes from the parent type scope. But I'm admittedly not an expert, therefore I'm not going to mess with what was already there.